04_documentation/code_documentation_tools_demo.md

```diff
@@ -0,0 +1,302 @@
+# Code Documentation Tools Demo
+
+## TODO
+
+- Tag the important commits in [python-code-documentation repository](https://github.com/Simulation-Software-Engineering/python-code-documentation) and replace the tag **TODO**s.
+    - This will be removed later when it is clear that the repository does not change anymore.
+
+## 1. Docstrings
+
+Goal: Add docstrings to the Python Code
+
+### Steps
+
+Initial state is in tag **TODO**
+
+1. Explain Python code shortly. Emphasize that we have a class now so we show how to document a class. We also moved all code in one python file to avoid problems with module imports.
+2. Add docstrings to class, solve, and auxiliary functions. Ideally add docstring only to one and then do fast-forward to tag **TODO** of repository.
+3. Show `__doc__` attribute and `help` function of Python. Inside the directory containing the `diffusionsolver.py` we start python and print the content of the docstrings.
+
+  ```bash
+  python
+  Python 3.8.10 (default, Sep 28 2021, 16:10:42)
+  [GCC 9.3.0] on linux
+  Type "help", "copyright", "credits" or "license" for more information.
+  >>> import diffusionsolver
+  >>> print(diffusionsolver.__doc__)
+
+  Solving the two-dimensional diffusion equation
+
+  Example acquired from https://scipython.com/book/chapter-7-matplotlib/examples/the-two-dimensional-diffusion-equation/
+
+  >>> help(diffusionsolver)
+
+      Some longer description is being generated
+
+  >>>
+  ```
+
+## 2. Sphinx
+
+We follow the [quickstart guide](https://www.sphinx-doc.org/en/master/usage/quickstart.html) and the [tutorial](https://www.sphinx-doc.org/en/master/tutorial/index.html) of Sphinx here.
+
+- Change back to the root of the repository
+- Reformat the `README.md` into `README.rst`
+
+    1. Rename the file
+
+       ```bash
+       git mv README.md README.rst
+       ```
+
+       We do this also for consistency with the new files that will be created.
+    2. Redo headlines by underlining them
+    3. Redo inline code samples with double backticks
+    4. Reformat code block using `code-block` directive
```

Comment:
Couldn't you do this with pandoc? :grin:

Reply:
I could, but 
- I do not want to add extra complexity by a tool the students might not know. I do not expect to have pandoc introduced here.
- I would use this as short introduction to reStructuredText because I do not expect to have time to talk about the corresponding chapter in/from the Markup Language slides.